src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -107,30 +100,42 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
      *            until the `Cargo.toml` becomes sound
      */
     override val projectDescription: CargoProjectDescription?
-        get() = lock (lock) { cached }
+        get() {
+            check(ApplicationManager.getApplication().isReadAccessAllowed)
+            return cached
+        }
 
-    private fun notifyCargoProjectUpdate(r: UpdateResult) {
+    private fun commitUpdate(r: UpdateResult) {
         ApplicationManager.getApplication().assertIsDispatchThread()
+        if (module.isDisposed) return
+
+        runWriteAction {
+            if (r is UpdateResult.Ok) {
+                cached = r.projectDescription
+                updateModuleDependencies(r.projectDescription);
+            }
+            notifyCargoProjectUpdate(r)
```

Comment:
>BTW, your listener should run out of dispatch-thread -- all UI related activity should go through EDT only.

Not sure I have understood you correctly here, can you please elaborate? 

Looks like root modification messages [are dispatched](https://upsource.jetbrains.com/idea-ce/file/idea-ce-c93c327673bdca7d0f0d022e82842485cd3bad59/platform/projectModel-impl/src/com/intellij/openapi/module/impl/ModuleManagerImpl.java?nav=31266:31350:focused&line=1028) in EDT ( an even in write action ??? )

In any case, we have only one listener whose job is to show a popup.